File: JZ/JZ6.从尾到头打印链表.py
# -*- coding: utf-8 -*-
# @Author  : Zhirong Tang
# @Time    : 2022/2/20 11:37 PM

# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
#
# @param listNode ListNode类
# @return int整型一维数组
#
class Solution:
    def __init__(self):
        self.res = list()
        
    def printListFromTailToHead(self , listNode: ListNode) -> List[int]:
        # write code here
        # method-1: force
        res = list()
        p = listNode
        if not p:
            return []
        while p:
            res.append(p.val)
            p = p.next
        return res[::-1]

        # method-2: 递归/DFS
        if listNode:
            self.printListFromTailToHead(listNode.next)
            self.res.append(listNode.val)
        return self.res
    
        # method-3: stack
        if not listNode:
            return []
        stack = list()
        res = []
        p = listNode
        while p:    # 进栈
            stack.append(p.val)
            p = p.next
        while stack:  # 出栈
            res.append(stack.pop())
        return res
        
        
        

File: JZ/JZ6.从尾到头打印链表.cpp
/*
 * @Description: 
 * @Author: Zhirong
 * @Date: 2022-02-21 15:10:47
 * @LastEditTime: 2022-02-21 15:48:23
 * @LastEditors: Zhirong
 */

/**
*  struct ListNode {
*        int val;
*        struct ListNode *next;
*        ListNode(int x) :
*              val(x), next(NULL) {
*        }
*  };
*/
class Solution {
public:
    vector<int> ans;
    vector<int> printListFromTailToHead(ListNode* head) {
        // method-1:暴力
        vector<int> ans;
        while (head) {
            ans.push_back(head->val);
            head = head->next;
        }
        std::reverse(ans.begin(), ans.end());
        return ans;

        // method-2:递归/DFS
        if (head) {
            printListFromTailToHead(head->next);
            ans.push_back(head->val);
        }
        return ans;

        // method-3:
        vector<int> ans;
        stack<int> stk;
        if (head == nullptr) {
            return ans;
        }
        while (head) {
            stk.push(head->val);
            head = head->next;
        }
        while (!stk.empty()) {
            int tmp = stk.top();
            stk.pop();
            ans.push_back(tmp);
        }
        return ans;
    }
};